### PR TITLE
Replace ortools with pulp

### PR DESCRIPTION
**Pull request by Charl1996**
## Technical Summary
[Ticket](https://dimagi-dev.atlassian.net/browse/SC-3163)

This PR simply replaces `ortools` with [pulp](https://pypi.org/project/PuLP/). The main reason is because ortools uses NumPy which SaaS seems less optimistic about from a maintenance perspective. This new package does not make use of numpy and is purely written in python. It's also quite a popular and widely used package.

The downside to using pulp instead of ortools is performance for optimizing case distribution amongst CHWs, but I would not too concerned about that since
1. We can simply put a loading screen on the UI if the disbursement is running if it's necessary (it's a feature that could be expected to take a while) 
2. Most users would probably not opt for the disbursement algorithm that does make use of pulp (radial solver), but rather the mapbox route solver (optimize over travel time on a road network)   

The method where we used an ortools solver was replaces by an equivalent solver of pulp.

## Feature Flag
Geospatial

### Safety story
There is an existing unit test that will test the method changed in this PR.

### QA Plan
No QA plan (yet).

### Rollback instructions
- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

### CODE REVIEW

**Comment by zandre-eng** [comment]
@Charl1996 I'm curious to hear if the results from this library were compared against ORTools? If so, were there any noticable differences in how cases were being grouped with this library compared to ORTools?

**Review by SmittieC — APPROVED** [review]
Looks fine if the results are approx the same

File: corehq/apps/geospatial/routing_solvers/pulp.py [hunk 1]
```diff
@@ -1,13 +1,13 @@
...
 import requests
+import pulp
 
-from ortools.linear_solver import pywraplp
 from django.conf import settings
 from .mapbox_optimize import validate_routing_request
 
 
-class ORToolsRadialDistanceSolver:
+class RadialDistanceSolver:
```
Comment (mkangia) [inline corehq/apps/geospatial/routing_solvers/pulp.py:10]: is this in use currently at all?

**Review by mkangia — COMMENTED** [review]
The requirements updates look okay.
I don't have context on changes to the tools usages itself.

File: corehq/apps/geospatial/routing_solvers/pulp.py (hunk 1 above)
Comment (Charl1996) [reply corehq/apps/geospatial/routing_solvers/pulp.py:10]: No. A ticket I have in progress makes use of this.

**Comment by Charl1996** [comment]
> @Charl1996 I'm curious to hear if the results from this library were compared against ORTools? If so, were there any noticable differences in how cases were being grouped with this library compared to ORTools?

No noticable differences. In fact, the example I gave it it grouped it exactly the same.

**Review by zandre-eng — APPROVED** [review]
Given that the results are the same, all good from my end.

**Comment by Charl1996** [comment]
Just an update: I saw a test failed where haversine could not find numpy. Turns out haversine does not depend on numpy, but they have a method that uses numpy (`haversine_verctor`) which was still used in our code. This particular test passed on my own environment because I had numpy still installed in my environment.
I changed the function a bit to use the non-vectorized function which only means we have to manually construct the matrix. This is actually [recommended](https://github.com/mapado/haversine/blob/8f3dc9f889a2cefac20838cf573141f5642daf88/haversine/haversine.py#L232C1-L234C84) if one does not have numpy installed.

The existing unit test still passes with the current state of the code in this PR and all the calculated values are precisely the same, so no change really except maybe a little performance on large datasets (haven't found a difference testing manually, but then again, it's hard to tell with a small dataset).

**Comment by Charl1996** [comment]
Sorry, I'm going to rerequest review for this last change.

**Review by zandre-eng — APPROVED** [review]
> so no change really except maybe a little performance on large datasets

We will generally be working with small datasets since disbursement will happen on a per-page basis. Given this, I think performance should be okay unless we start increasing what page size the user may have.

Good catch!

**Pull request merged**